### PR TITLE
Added integration with Mod Menu

### PR DESCRIPTION
**Pull request by poqbox**
I've only done a bit of modding with Fabric, but I wanted to be able to configure this mod through Mod Menu. I also saw that this was planned in #30. I haven't modded with NeoForge, so I'm not sure if it will work with that.

There are a couple issues I noticed, so I'll list them here.
- The Overlay Settings screen with the sliders doesn't render properly when on the Title screen, but is fine in-game.
- I removed the option to open the config screen with Fn+7. I had to change ModeSwitcherScreen.java a bit to get it to work with Mod Menu, but this created an error with KeyBind.java and I didn't know how to adapt it.

This should work without any config mods like Cloth config or YACL.

### CODE REVIEW

**Comment by poqbox** [comment]
I didn't changed the mod version number either. Wasn't sure of which number to change.

**Comment by andi-makes** [comment]
Thank you very much for this PR! I appreciate your contribution!

I am currently quite busy with university assignments and adding finishing touches for Lighty for Minecraft 1.20.5, but I hope to be able to review this PR and #58 this weekend. Thank you for your understanding!

File: common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java [hunk 1]
```diff
@@ -48,9 +48,6 @@ public static void handleKeyBind(Minecraft client) {
         if (newToggleState && !KeyBind.oldToggleState) {
             ModeLoader.toggle();
         }
-        if (newKeyState && !KeyBind.oldKeyState) {
```
Comment (andi-makes) [inline common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]: Being able to change the Lighty Settings is pretty important, please revert this removal.

File: common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java [hunk 2]
```diff
@@ -112,11 +112,11 @@ protected void init() {
     }
 
     @Override
-    public void render(@NotNull GuiGraphics guiGraphics, int i, int j, float f) {
+    public void render(@NotNull GuiGraphics guiGraphics, int mouseX, int mouseY, float delta) {
```
Comment (andi-makes) [inline common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java]: I know that this is a good intended change, but please only change the things you need in your PR.  Changing unrelated things may make porting changes more difficult than it already is.

File: fabric/build.gradle [hunk 3]
```diff
@@ -35,6 +36,11 @@ dependencies {
     // Fabric API. This is technically optional, but you probably want it anyway.
     modImplementation "net.fabricmc.fabric-api:fabric-api:${rootProject.findProperty("fabric_api_version")}"
 
+    // Mod Menu
+    modImplementation "com.terraformersmc:modmenu:${rootProject.findProperty("modmenu_version")}", {
+        exclude module: 'fabric-api'
```
Comment (andi-makes) [inline fabric/build.gradle:41]: I am curious: is there a benefit to excluding the fapi module from modmenu? It seems like this isn't strictly needed based on me looking at some other projects

File: fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package dev.schmarrn.lighty.fabric;
+
+import com.terraformersmc.modmenu.api.ConfigScreenFactory;
+import com.terraformersmc.modmenu.api.ModMenuApi;
+import dev.schmarrn.lighty.ui.ModeSwitcherScreen;
+
+public class ModMenuIntegration implements ModMenuApi {
+    @Override
+    public ConfigScreenFactory<?> getModConfigScreenFactory() {
+        return ModeSwitcherScreen::new;
```
Comment (andi-makes) [inline fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java]: It doesn't make sense to show the entire ModeSwitcherScreen, just showing the SettingsScreen is enough.

By showing the entire ModeSwitcherScreen, users can click on the "Overlay: OFF" Button, but unexpected things happen because there is no world loaded.

File: common/src/main/java/dev/schmarrn/lighty/ui/ModeSwitcherScreen.java
Comment (andi-makes) [inline common/src/main/java/dev/schmarrn/lighty/ui/ModeSwitcherScreen.java:1]: You can revert this change and instead implement a second, additional Constructor for SettingsScreen.

By taking in the parent Screen, you also need to override the onClose method as follows:

```java
@Override
public void onClose() {
    assert this.minecraft != null;
    this.minecraft.setScreen(parent);
}
```

That way, by closing the Lighty Settings Menu the user returns to the Modmenu Screen.

**Review by andi-makes — CHANGES_REQUESTED** [review]
Sorry to have kept you waiting, but at last here is my review!

First of, thanks again for working on Lighty, it is highly appreciated!

There are a few minor issues/questions I have regarding this PR, see the comments.

Additionally, the Settings Screen does seem to be broken when opened, see the attached screenshot below. The sliders are half-missing. I currently have no idea what the cause for that is, but this needs to be fixed before this PR can get merged.

image

File: common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java (hunk 1 above)
Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]: Sorry about that. Because of how I implemented the integration in ModMenuIntegration.java, I needed to change the constructor for ModeSwitcherScreen(), which caused an error with the keybind. At the time, I removed the keybind to get rid of the error, thinking that ModMenu would be a good enough substituion.

However, you're right in being able to change the settings being important, since some might not install ModMenu with it.

I'll add a second constructor to fix this. Doing so should restore the original keybind, while keeping integration with ModMenu.

File: common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java (hunk 2 above)
Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java]: Sorry about this one. It was a change I made after the pull request. I didn't know that it would be added to the same PR. 

I'll make a new PR for this.

File: fabric/build.gradle (hunk 3 above)
Comment (poqbox) [reply fabric/build.gradle:41]: I'm not entirely sure, but I would imagine that it cuts down on the amount of things that gradle is fetching. It's a practice that I saw other modders doing, so I figured that I'd follow it. If this is the case, I would assume that there's no harm in it.

I've never tested it, though.

File: fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java (hunk 4 above)
Comment (poqbox) [reply fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java]: I had thought this was by design, so that they can switch between the various modes. Wouldn't the SettingsScreen only bring up the screen with the slider? My experience with Lighty V2.2.0 is that F7 would bring up the "Lighty Mode Switcher" screen, and provide an option to bring up the SettingsScreen with the "Overlay Settings" button at the bottom.

I think being able to change Lighty modes via the settings screen is a good design choice, and should be kept. If you want users to start on the SettingsScreen, I think combining the various modes into one widget/button and moving it to the SettingsScreen would help to unify the screens into one without cluttering it. The new button would act like the existing "Graphics" or "GUI Scale" button, where clicking it would cycle between the various modes.

File: common/src/main/java/dev/schmarrn/lighty/ui/ModeSwitcherScreen.java
Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/ui/ModeSwitcherScreen.java:1]: Good point, will do.

**Comment by poqbox** [comment]
I've made some changes.

No need to rush the review if you're still busy, I just did it cause I felt like it.

File: common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java [hunk 5]
```diff
@@ -121,8 +127,8 @@ public void render(@NotNull GuiGraphics guiGraphics, int i, int j, float f) {
 
     @Override
     public void onClose() {
-        Compute.clear();
-        super.onClose();
+        assert this.minecraft != null;
```
Comment (andi-makes) [inline common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java:131]: The `Compute.clear()` call is important to reset the overlay after changing the settings. It *must* be called when closing the settings menu.

```suggestion
        Compute.clear();
        assert this.minecraft != null;
```

File: fabric/build.gradle (hunk 3 above)
Comment (andi-makes) [reply fabric/build.gradle:41]: That's fair enough, it does seem to work after all.

File: fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java (hunk 4 above)
Comment (andi-makes) [reply fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java]: With how the Lighty Menu is designed in 1.20.3/4 and because of the inner workings, I personally strongly believe that only showing the SettingsScreen and not the ModeSwitcherScreen is the right way to do it.

With 1.20.5, I refactored (or, am still refactoring) the UI parts of Lighty. The design in 1.20.5 is better suited for what you are envisioning.

I suggest this: For 1.20.3/4, we only show the SettingsScreen. When I port this PR to 1.20.5, I will make the required changes to also let the user change the default mode using ModMenu.

I hope that's a good compromise. I really don't want to backport the changes I made for 1.20.5, because I changed a lot and it would be a lot of work :upside_down_face:

**Review by andi-makes — CHANGES_REQUESTED** [review]
It already looks good, and there are only a few minor things to iron out.

I've found a fix for the sliders not showing when being in ModMenu, I'll commit that after this PR is merged.

Thanks again for the work!

File: fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java (hunk 4 above)
Comment (poqbox) [reply fabric/src/main/java/dev/schmarrn/lighty/fabric/ModMenuIntegration.java]: Alright, it's been done.

File: common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java (hunk 5 above)
Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/ui/SettingsScreen.java:131]: It's been restored.

**Comment by poqbox** [comment]
No problem, just glad to get ModMenu integration with Lighty.

File: common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ public static void handleKeyBind(Minecraft client) {
             ModeLoader.toggle();
         }
         if (newKeyState && !KeyBind.oldKeyState) {
-            client.setScreen(new ModeSwitcherScreen());
+            client.setScreen(new SettingsScreen());
```
Comment (andi-makes) [inline common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]:
```suggestion
            client.setScreen(new ModeSwitcherScreen());
```

This has to be ModeSwitcherScreen, not SettingsScreen

File: fabric/build.gradle
Comment (andi-makes) [inline fabric/build.gradle:1]: Please do not touch build files other than the changes you need to do in order for your feature to work. In this case: do not update Loom, Gradle, Fabric Api, Neoforge or Minecraft in this PR.

1.20.3 is compatible with 1.20.4, and because this branch is for both Minecraft versions, we need to leave fabric, loom etc. at 1.20.3.

To test if your PR works in 1.20.4, you'd generally test it outside of the dev environment.

Please revert the related changes, thanks!

File: gradle.properties
Comment (andi-makes) [inline gradle.properties:1]: See my comment on file `fabric/build.gradle`

File: gradle/wrapper/gradle-wrapper.properties
Comment (andi-makes) [inline gradle/wrapper/gradle-wrapper.properties:1]: See my comment on file `fabric/build.gradle`

**Review by andi-makes — CHANGES_REQUESTED** [review]
Sorry to have kept you waiting again!

Only a few more changes, please see the comments for further info.

File: common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java (hunk 6 above)
Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]: Are you sure? Right now, both this and the ModMenu settings start on the SettingsScreen. Changing this would ruin the parity between the two.

File: fabric/build.gradle
Comment (poqbox) [reply fabric/build.gradle:1]: Understood, I'll revert it.

File: gradle.properties
Comment (poqbox) [reply gradle.properties:1]: Got it.

File: gradle/wrapper/gradle-wrapper.properties
Comment (poqbox) [reply gradle/wrapper/gradle-wrapper.properties:1]: Yep.

File: common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java (hunk 6 above)
Comment (andi-makes) [reply common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]: Yes, I am sure, the ModMenu settings screen should be SettingsScreen, while the Keybind still opens ModeSwitcherScreen.

Eventually in 1.20.5/1.20.6 after porting this PR to those versions both will point to the same thing, but in 1.20.3/1.20.4 I'd prefer if they are still different because of internal technical reasons.

Comment (poqbox) [reply common/src/main/java/dev/schmarrn/lighty/event/KeyBind.java]: Alright, it's been done.

**Review by andi-makes — APPROVED** [review]
Looks good to me!

Thank you for your work, it is highly appreciated! I'll try to release this either this evening or tomorrow, thanks again!

**Comment by poqbox** [comment]
np
